lib/src/api/opt/auth.rs

```diff
@@ -12,6 +13,108 @@ pub struct Signup;
 #[derive(Debug)]
 pub struct Signin;
 
+/// Credentials level
+#[derive(Debug, Clone)]
+pub enum CredentialsLevel {
+	Root,
+	Namespace,
+	Database,
+	Scope,
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum Error {
+	#[error("Username is needed for authentication but it was not provided")]
+	NoUsername,
+	#[error("Password is needed for authentication but it was not provided")]
+	NoPassword,
+	#[error("Namespace is needed for authentication but it was not provided")]
+	NoNamespace,
+	#[error("Database is needed for authentication but it was not provided")]
+	NoDatabase,
+	#[error("Scope is needed for authentication but it was not provided")]
+	NoScope,
+	#[error("Scope params are needed for authentication but they were not provided")]
+	NoScopeParams,
+}
+
+/// Construct a Credentials instance for the given auth level
+#[derive(Debug, Default)]
+pub struct CredentialsBuilder<'a> {
```

Comment:
All these new additions in this module make sense to me from the CLI not the SDK. Is there any reason these are in the SDK? For example, I don't think using a builder here is an improvement to using the structs directly like we already do. On the contrary, it now makes it possible to make mistakes that were not possible before, hence we now need to introduce a new error type.

Reply:
Thank you! I have moved these types to the CLI in [526a849](https://github.com/surrealdb/surrealdb/pull/2985/commits/526a849fa33e5194ac207ffc69d2ead389e98c28).